app/jobs/create_virtual_objects_job.rb

```diff
@@ -22,85 +20,50 @@ def perform(bulk_action_id, params)
 
     virtual_objects = VirtualObjectsCsvConverter.convert(csv_string: params[:create_virtual_objects])
 
-    # rubocop:disable Metrics/BlockLength
     with_bulk_action_log do |log|
       log.puts("#{Time.current} Starting #{self.class} for BulkAction #{bulk_action_id}")
 
-      parent_druids = parent_ids_from(virtual_objects)
-
       # NOTE: We use this instead of `update_druid_count` because virtual object
       #       creation does not use the `pids` form field.
-      bulk_action.update(druid_count_total: parent_druids.length)
-
-      problematic_druids = find_problematic_druids(parent_druids)
-
-      if problematic_druids.values.flatten.any?
-        log.puts("#{Time.current} #{UNAUTHORIZED_MESSAGE}: #{problematic_druids[:unauthorized].to_sentence}") if problematic_druids[:unauthorized].any?
-        log.puts("#{Time.current} #{NOT_FOUND_MESSAGE}: #{problematic_druids[:not_found].to_sentence}") if problematic_druids[:not_found].any?
-        bulk_action.increment!(:druid_count_fail, problematic_druids.values.flatten.length)
-        # Remove problematic parent druids from list of objects to be operated upon
-        virtual_objects[:virtual_objects].reject! do |virtual_object|
-          problematic_druids.values.flatten.include?(virtual_object[:parent_id])
+      bulk_action.update(druid_count_total: virtual_objects.length)
+
+      # NOTE: `ability` is defined in this job's superclass, `GenericJob`
+      not_found_druids, unauthorized_druids = ProblematicDruidFinder.find(druids: parent_ids_from(virtual_objects), ability: ability)
+      problematic_druids = not_found_druids + unauthorized_druids
+
+      if problematic_druids.any?
+        log.puts("#{Time.current} #{UNAUTHORIZED_MESSAGE}: #{unauthorized_druids.to_sentence}") if unauthorized_druids.any?
+        log.puts("#{Time.current} #{NOT_FOUND_MESSAGE}: #{not_found_druids.to_sentence}") if not_found_druids.any?
+        bulk_action.increment!(:druid_count_fail, problematic_druids.length)
+
+        # Short-circuit if all parent druids were problematic
+        if problematic_druids.length == virtual_objects.length
+          log.puts("#{Time.current} No virtual objects could be created. See other log entries for more detail.")
+          log.puts("#{Time.current} Finished #{self.class} for BulkAction #{bulk_action_id}")
+          return
```

Comment:
What the heck???

Reply:
100% heck.

Reply:
![](https://media.giphy.com/media/v2OWXglWr9PAk/giphy-downsized.gif)